testnet/stacks-node/src/nakamoto_node/relayer.rs

```diff
@@ -880,29 +905,79 @@ impl RelayerThread {
         };
         let mining_pkh = Hash160::from_node_public_key(&StacksPublicKey::from_private(mining_key));
 
-        // If we won the last sortition, then we should start a new tenure off of it.
-        let last_block_election_snapshot = {
+        let highest_tenure_start_block_header = NakamotoChainState::get_tenure_start_block_header(
+            &mut self.chainstate.index_conn(),
+            &canonical_stacks_tip,
+            &canonical_stacks_tip_ch,
+        )
+        .map_err(|e| {
+            error!(
+                "Relayer: Failed to get tenure-start block header for stacks tip {canonical_stacks_tip}: {e:?}"
+            );
+            NakamotoNodeError::ParentNotFound
+        })?
+        .ok_or_else(|| {
+            error!(
+                "Relayer: Failed to find tenure-start block header for stacks tip {canonical_stacks_tip}"
+            );
+            NakamotoNodeError::ParentNotFound
+        })?;
+        let highest_tenure_bhh =
+            BlockHeaderHash(highest_tenure_start_block_header.index_block_hash().0);
+
+        // If we won the last good sortition, then we should extend off of it.
+        let last_good_block_election_snapshot = {
             let ih = self.sortdb.index_handle(&burn_tip.sortition_id);
-            ih.get_last_snapshot_with_sortition(burn_tip.block_height)
+            info!(
+                "Relayer: Getting last snapshot with sortition for {}",
+                burn_tip.block_height
+            );
+
+            let sn = ih
+                .get_last_snapshot_with_sortition(burn_tip.block_height)
                 .map_err(|e| {
                     error!("Relayer: failed to get last snapshot with sortition: {e:?}");
                     NakamotoNodeError::SnapshotNotFoundForChainTip
+                })?;
+            if sn.winning_stacks_block_hash != highest_tenure_bhh {
+                info!(
+                    "Relayer: Sortition winner is not committed to the canonical tip; allowing last miner to extend";
+                    "burn_block_height" => burn_tip.block_height,
+                    "consensus_hash" => %burn_tip.consensus_hash,
+                );
+
+                SortitionDB::get_block_snapshot_consensus(
```

Comment:
At line 984 we run this exact query. With your change, is that always necessary? If it's maybe necessary, can we change 984 to something like:

```rust
let canonical_block_snapshot = if last_good_block_election_snapshot
    .canonical_stacks_tip_consensus_hash
    == canonical_stacks_tip_ch
{
    last_good_block_election_snapshot
} else {
    SortitionDB::get_block_snapshot_consensus(self.sortdb.conn(), &canonical_stacks_tip_ch)
        .map_err(|e| {
            error!("Relayer: failed to get block snapshot for canonical tip: {e:?}");
            NakamotoNodeError::SnapshotNotFoundForChainTip
        })?
        .ok_or_else(|| {
            error!("Relayer: failed to get block snapshot for canonical tip");
            NakamotoNodeError::SnapshotNotFoundForChainTip
        })?
};
```

Reply:
I do think it's still necessary. We can definitely reuse it here and avoid looking it up again.

Reply:
Hmm... maybe it's not. I'll look at this some more.

Reply:
IMO I would appreciate some ELI5(ish) comments on some of the sortdb lookups to explain why/what they're doing. There's already great comments in here, I think it's just tricky to follow when I have to constantly look up with a specific function does

Reply:
Ok, I tried to clean this function up a bit. What do you think about this refactoring? 21788465f

Reply:
I also improved the names and added more comments in another commit just now.